enclave_build/src/image_manager.rs

```diff
@@ -0,0 +1,180 @@
+                Err(_) => None,
+            };
+
+        let image_details = Self::fetch_image_details(&image_name, storage).await?;
+
+        Ok(Self {
+            image_name,
+            image_details,
+        })
+    }
+
+    /// Returns a struct containing image metadata.
+    ///
+    /// If the image is stored correctly, the function tries to fetch the image from the storage.
+    ///
+    /// If the image is not stored or a storage was not created (the 'storage' field is None),
+    /// it pulls the image, stores it (if the 'storage' field is not None) and returns its metadata.
+    ///
+    /// If the pull succeeded but the store operation failed, it returns the pulled image metadata.
+    async fn fetch_image_details(
+        image_name: &str,
+        mut storage: Option<OciStorage>,
+    ) -> Result<ImageDetails> {
+        let local_storage = storage.as_mut();
+
+        let image_details = if let Some(storage) = local_storage {
+            // Try to fetch the image from the storage
+            storage.fetch_image_details(image_name).map_err(|err| {
+                // Log the fetching error
+                eprintln!("{:?}", err);
+                err
+            })
+        } else {
+            Err(EnclaveBuildError::OciStorageNotFound(
+                "Local storage missing".to_string(),
+            ))
+        };
+
+        // If the fetching failed, pull it from remote and store it
+        match image_details {
+            Ok(details) => Ok(details),
+            Err(_) => {
+                // The image is not stored, so try to pull and then store it
+                let image_data = crate::pull::pull_image_data(image_name).await?;
+
+                // If the store operation fails, discard error and proceed with getting the details
+                if let Some(local_storage) = storage.as_mut() {
+                    local_storage
+                        .store_image_data(image_name, &image_data)
+                        .map_err(|err| eprintln!("Failed to store image: {:?}", err))
+                        .ok();
+                }
+
+                // Get the image metadata from the pulled struct
+                ImageDetails::build_details(image_name, &image_data)
+            }
+        }
+    }
+}
+
+/// Adds the default ":latest" tag to an image if it is untagged
+fn normalize_tag(image_name: &str) -> Result<String> {
+    let image_ref = Reference::try_from(image_name).map_err(|err| {
+        EnclaveBuildError::ImageBuildError(format!("Invalid image name format: {}", err))
+    })?;
+
+    match image_ref.tag() {
+        Some(_) => Ok(image_name.to_string()),
+        None => Ok(format!("{}:latest", image_name)),
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use sha2::Digest;
+
+    use super::{normalize_tag, OciImageManager};
+
+    #[cfg(target_arch = "x86_64")]
+    const SAMPLE_IMAGE: &str =
+        "667861386598.dkr.ecr.us-east-1.amazonaws.com/enclaves-samples:vsock-sample-server-x86_64";
```

Comment:
Let's publish some empty test images into our Public ECR so that anyone can pull these instead.